src/vision/status_quo/grace_wants_to_integrate_c_api.md

```diff
@@ -0,0 +1,176 @@
+
+### The first problem: polls and wake-ups
+
+Grace sets out to start writing the pipeline to get frames from the cameras. She realizes that while the polling call
+that the manufacturer provided in their library is similar in nature to a future, it doesn't quite encompass everything.
+In C, one might have to set some kind of heartbeat timer for polling. Grace explains to her team that this heartbeat is
+similar to how the `Waker` object works in a `Future`'s `Context` type, in that it is how often the execution
+environment should re-try the future if the call to `poll` returns `Poll::Pending`.
+
+Something Grace learns to love immediately about Rust is that writing the `Future` here does not require her team to
+write their own execution environment. In fact, the future can be entirely written independent of the execution
+environment. She quickly writes an async method to represent the polling process:
+
+```rust
+/// Gets the next frame from the camera, waiting `retry_after` time until polling again if it fails.
+///
+/// Returns Some(frame) if a frame is found, or None if the camera is disconnected or goes down before a frame is
+/// available.
+async fn next_frame(camera: &Camera, retry_after: Duration) -> Option<Frame> {
+    while camera.is_available() {
+        if let Some(frame) = camera.poll() {
+            return Some(frame);
+        } else {
+            task::sleep_for(retry_after).await;
+        }
+    }
+
+    None
+}
+```
+
+The underlying C API doesn't provide any hooks that can be used to wake the `Waker` object on this future up, so Grace
+and her team decide that it is probably best if they just choose a sufficiently balanced `retry_after` period in which
+to try again. It does feel somewhat unsatisfying, as calling `sleep_for` feels about as hacky as calling
+`std::this_thread::sleep_for` in C++. However, there is no way to directly interoperate with the waker without having a
+separate thread of execution wake it up, and the underlying C library doesn't have any interface offering a notification
+for when that should be. In the end, this is the same kind of code that they would write in C, just without having to
+implement a custom execution loop themselves, so the team decides it is not a total loss.
+
+### The second problem: doing this many times
+
+Doing this a single time is fine, but an end goal of the project is to be able to stream frames from the camera for
+unspecified lengths of time. Grace spends some time searching, and realizes that what she actually wants is a `Stream`
+of some kind, so that her team can write something similar to:
+
+```rust
+let frame_stream = stream_from_camera(camera, Duration::from_millis(5));
+
+while let Some(frame) = frame_stream.next().await {
+    // process frames
+}
+
+println!("Frame stream closed.");
+```
+
+She scours existing crates, in particular looking for one way to transform the above future into a stream that can be
```

Comment:
The use of the word "scours" makes me think this wasn't a trivial or pleasant process. Perhaps another moral is that understanding that this use case is best modeled as a stream *and* learning how to turn futures into streams were both not easy for someone with a C/C++ background. What do you think?

Reply:
Good catch -- I would like to hear a bit more about how she figured out that she wanted streams and found the combinator, because that's not entirely obvious to me either.

Reply:
I think the main difficulty (why I used the word "scours") was that it was not easy to search based off of types. I was originally looking into how to implement the Stream trait while using async calls inside `poll_next`! I have a background that also extends to using streams in Scheme, so I was thinking there about how I'd do it, and that was when I started looking through `futures::stream` for possible combinators or constructors for stream types. 

Eventually after reading through the whole module I found the one combinator that had an example taking in a future, instead of constructing a stream from synchronous code. I didn't put 2 and 2 together with regards to `stream::unfold` in Rust and `stream-unfold` in Scheme (see SRFI 45) until I had already read over the code 😅 

But I would generally agree again with this point:

> Perhaps another moral is that understanding that this use case is best modeled as a stream and learning how to turn futures into streams were both not easy for someone with a C/C++ background. What do you think?

Shall I add it to the morals section?

Reply:
Added a note to the morals section in 9e0bd1c5f1a95b732361ddcb898d2cd0decb501c. 

Let me know if that clears it up!

Reply:
It looks like this is resolved.